tools/autograd/gen_variable_type.py

```diff
@@ -355,6 +355,50 @@ def lookup_pred(pred, xs):
         """Return the index of the first element of xs matching pred."""
         return next((i, x) for i, x in enumerate(xs) if pred(x))
 
+    def is_nn_fwd(defn_name, declarations_by_name):
```

Comment:
Are these still necessary: `is_nn_fwd` and `preprocess_nn_function`?

Reply:
Yes, at the moment. `nn` functions are treated in a special way, because we have to dispatch on the `_forward` functions that take buffers as arguments. These are not explicitly declared anywhere, we just know they're generated by `ATen/nn_parse.py` for NN, non-backward functions.

We can change this by having a `dispatch` property somewhere in a yaml file in ATen (like it's done for `ATen/cudnn`). I took the simpler route following our previous exchange on not changing strategy for `nn_parse`.

Reply:
They are explicitly declared: when I look in Declarations.yaml I see avg_pool2d_forward and get function stubs in Function.h.

The root cause of this problem is the ugly buffers design. In some cases THNN could be refactored to eliminate them (since we have a caching allocator now) but this doesn't apply universally. Perhaps we can annotate the buffers in the argument list in `derivatives.yaml` (so it doesn't look like the variables in backwards pop out of no where), but with some sort of syntax like `foo=BUFFER` making it clear that we will generate an API that doesn't take this buffer. A more parsimonious mechanism to deal with buffers would be to return the buffers as legit output tuples, but then you need another mechanism to not *actually* return the buffers from user facing code. Today's design of arguments seems easier.

Reply:
Yes, they are declared, but the declarations are generated by `parse_nn.py` and there’s currently no way of telling that `nll_loss` dispatches `nll_loss_forward`, there's no explicit link between them except their name.
We could get rid of `is_nn_fwd` and `preprocess_nn_function` by introducing a `dispatch` property in `Declarations.yaml` and instructing `parse_nn` to generate it. What do you think?

As for buffers, we know they are there in the generator, but I agree it’s ugly that they pop out of nowhere in backwards. Maybe there’s another option to annotate `derivatives.yaml`: do it after the function signature, something like:
```
- name: nll_loss(...args...) buffers=[foo, bar]
```
or
```
- name: nll_loss(...args...) [foo, bar]
```
The alternative is have a reserved key for buffers, or refactor the schema in `derivatives.yaml` to cleanly allow keys for annotating signatures.

Reply:
Yes, you are right.

Thinking about this more, buffers seem to introduce a staging problem:

- In Tensor, we would like foo and foo_forward, because a user who only plans to do inference prefers not to have to create the buffers themselves.

- In Variable, we'd also like foo and foo_forward, but we cannot directly forward foo to foo in ATen; instead, it must be forwarded to foo_forward, so that we have access to the buffers to save for backwards.

So, coincidentally, Tensor foo and Variable foo have the same arguments, but they must be implemented twice because the Variable version must dispatch to Tensor foo_forward, NOT Variable foo_forward, as the native mechanism would ordinarily do (there is NO reason why you would expose such a thing.)

I'm satisfied with the current mechanism, and OK if we just document the "buffer out of nowhere" behavior.